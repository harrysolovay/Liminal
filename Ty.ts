import type { Ref } from "./Ref.ts"

export function Ty<T, P extends string = never>(
  toSchema: ToSchema,
  descriptions: Description[] = [],
  applied: Applied = {},
): Ty<T, P> {
  return Object.assign(
    <P2 extends string[]>(template: TemplateStringsArray, ...placeheld: P2) =>
      Ty<T, P | P2[number]>(toSchema, [{ template, placeheld }, ...descriptions], applied),
    {} as { [Ty.T]: T; [Ty.P]: P },
    {
      toSchema,
      descriptions,
      applied,
      apply: <A extends Partial<Record<P, string | number>>>(values: A) => {
        return Ty<T, Exclude<P, keyof A>>(toSchema, descriptions, { ...applied, ...values })
      },
    },
  )
}

export interface Ty<T = any, P extends string = string> {
  <P2 extends string[]>(template: TemplateStringsArray, ...placeheld: P2): Ty<T, P | P2[number]>
  [Ty.T]: T
  [Ty.P]: P
  toSchema: ToSchema
  descriptions: Description[]
  applied: Applied
  apply: <A extends Partial<Record<P, number | string>>>(values: A) => Ty<T, Exclude<P, keyof A>>
}

export namespace Ty {
  export type T = typeof T
  export declare const T: unique symbol
  export type P = typeof P
  export declare const P: unique symbol
}

export type ToSchema = (description: string | undefined, ref: Ref) => Schema
export type Schema = Record<string, unknown>

export interface Description {
  template: TemplateStringsArray
  placeheld: string[]
}

export type Applied = Record<string, number | string>
