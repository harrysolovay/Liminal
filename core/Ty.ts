import type { Ref } from "./Ref.ts"
import { phantoms } from "../util/phantoms.ts"

export function Ty<T, P extends string = never>(
  toSchema: ToSchema,
  descriptions: Array<Description> = [],
  applied: Applied = {},
): Ty<T, P> {
  return Object.assign(
    <P2 extends Array<string>>(template: TemplateStringsArray, ...placeheld: P2) =>
      Ty<T, P | P2[number]>(toSchema, [{ template, placeheld }, ...descriptions], applied),
    phantoms<{ [Ty.T]: T; [Ty.P]: P }>(),
    {
      toSchema,
      descriptions,
      applied,
      apply: <A extends Partial<Record<P, string | number>>>(values: A) => {
        return Ty<T, Exclude<P, keyof A>>(toSchema, descriptions, { ...applied, ...values })
      },
    },
  )
}

export interface Ty<T = any, P extends string = string> {
  <P2 extends Array<string>>(
    template: TemplateStringsArray,
    ...placeheld: P2
  ): Ty<T, P | P2[number]>
  [Ty.T]: T
  [Ty.P]: P
  toSchema: ToSchema
  descriptions: Array<Description>
  applied: Applied
  apply: <A extends Partial<Record<P, number | string>>>(values: A) => Ty<T, Exclude<P, keyof A>>
}

export namespace Ty {
  export type T = typeof T
  export declare const T: unique symbol
  export type P = typeof P
  export declare const P: unique symbol
}

export type ToSchema = (description: string | undefined, ref: Ref) => Schema
export type Schema = Record<string, unknown>

export interface Description {
  template: TemplateStringsArray
  placeheld: Array<string>
}

export type Applied = Record<string, number | string>
