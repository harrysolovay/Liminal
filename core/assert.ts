import { assert as assert_, assertArrayIncludes, assertEquals } from "@std/assert"
import type { PartialType, Type } from "./Type.ts"
import { TypeVisitor } from "./TypeVisitor.ts"

export interface Diagnostic {
  type: PartialType
  description?: string
  path: string
  value: unknown
  exception?: unknown
}

export namespace Diagnostic {
  export function toString({ description, exception, path, value }: Diagnostic): string {
    return `Assertion${description ? ` "${description}"` : ""} failed ${
      exception
        ? exception instanceof Error ? `with Error "${exception.name}"` : "with exception"
        : ""
    } on value \`${JSON.stringify(value)}\` at \`root${path}\`: ${
      exception instanceof Error ? exception.message : JSON.stringify(exception)
    }`
  }
}

export async function assert(type: Type<any>, value: unknown): Promise<void> {
  const ctx = new AssertContext([], [], "root", value)
  visit(ctx, type)
  const diagnostics = [
    ...ctx.structuralDiagnostics,
    ...await Promise.all(ctx.annotationDiagnostics ?? []).then((v) => v.filter((e) => !!e)),
  ]
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}

export function matchUnionMember(members: Array<PartialType>, value: unknown): PartialType | void {
  const queue = [...members]
  while (queue.length) {
    const current = queue.pop()!
    const diagnostics: Array<Diagnostic> = []
    visit(new AssertContext(diagnostics, undefined, "", value), current)
    if (diagnostics.length) {
      continue
    }
    return current
  }
}

export class AssertContext {
  constructor(
    readonly structuralDiagnostics: Array<Diagnostic>,
    readonly annotationDiagnostics: undefined | Array<Promise<Diagnostic | void>>,
    readonly path: string,
    readonly value: unknown,
    readonly junction?: number | string,
  ) {}

  descend = (value: unknown, junction?: number | string): AssertContext =>
    new AssertContext(
      this.structuralDiagnostics,
      this.annotationDiagnostics,
      this.path,
      value,
      junction,
    )
}

const visit = TypeVisitor<AssertContext, void>({
  hook(next, ctx, type) {
    const path = ctx.junction ? ctx.path : `${ctx.path}["${ctx.junction}"]`
    try {
      next(ctx, type)
    } catch (exception: unknown) {
      ctx.structuralDiagnostics.push({
        type,
        path,
        exception,
        value: ctx.value,
      })
    }
    const { annotationDiagnostics } = ctx
    if (annotationDiagnostics) {
      type.annotations
        .filter((annotation) => typeof annotation === "object" && annotation?.type === "Assertion")
        .forEach(({ f, description }) => {
          if (f) {
            annotationDiagnostics.push((async () => {
              try {
                const passed = await f(ctx.value)
                if (!passed) {
                  return {
                    type,
                    path,
                    description,
                    value: ctx.value,
                  }
                }
              } catch (exception: unknown) {
                return {
                  type,
                  path,
                  exception,
                  description,
                  value: ctx.value,
                }
              }
            })())
          }
        })
    }
  },
  null({ value }) {
    assert_(value === null)
  },
  boolean({ value }) {
    assert_(typeof value === "boolean")
  },
  integer({ value }) {
    assert_(Number.isInteger(value))
  },
  number({ value }) {
    assert_(typeof value === "number")
  },
  string({ value }) {
    assert_(typeof value === "string")
  },
  const({ value }, _1, _2, cmpValue) {
    assertEquals(value, cmpValue)
  },
  enum({ value }, _1, ...values) {
    assert_(typeof value === "string")
    assertArrayIncludes(values, [value])
  },
  array(ctx, _1, element) {
    assert_(Array.isArray(ctx.value))
    ctx.value.forEach((value, i) => visit(ctx.descend(value, i), element))
  },
  object(ctx, _1, fields) {
    const keys = Object.keys(fields).toSorted()
    const { value } = ctx
    assert_(typeof value === "object" && value !== null)
    assertEquals(keys, Object.keys(value).toSorted())
    keys.forEach((key) => visit(ctx.descend(value[key as never], key), fields[key]!))
  },
  transform(ctx, _1, from) {
    visit(ctx, from)
  },
  union(ctx, _1, ...members) {
    const match = matchUnionMember(members, ctx.value)
    assert_(match)
    visit(ctx, match)
  },
  ref(ctx, _1, get) {
    visit(ctx, get())
  },
})
