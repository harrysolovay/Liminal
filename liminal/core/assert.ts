import { AssertionContext } from "./AssertionContext.ts"
import type { AnyType } from "./Type.ts"

export async function assert(this: AnyType, value: unknown): Promise<void> {
  const ctx = new AssertionContext("root", [], [])
  ctx.visit(this, value)
  const diagnostics = [
    ...ctx.structuralDiagnostics,
    ...await Promise
      .all(ctx.annotationDiagnostics ?? []).then((v) => v.filter((e) => !!e)),
  ]
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}
