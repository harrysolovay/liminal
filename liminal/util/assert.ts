import { LiminalAssertionError } from "../LiminalAssertionError.ts"

export function assert(expr: unknown, msg = ""): asserts expr {
  if (!expr) {
    throw new LiminalAssertionError(msg)
  }
}
