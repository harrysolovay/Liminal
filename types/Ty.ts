import { phantoms } from "../util/phantoms.ts"
import type { Ref } from "./Ref.ts"
import type { RootTy } from "./RootTy.ts"

export function Ty<T, P extends keyof any = never>(
  toSchema: ToSchema,
  context: Array<Context> = [],
  applied: Applied = {},
): Ty<T, P> {
  return Object.assign(
    <P2 extends Array<keyof any>>(template: TemplateStringsArray, ...placeheld: P2) =>
      Ty<T, P | P2[number]>(toSchema, [{ template, placeheld }, ...context], applied),
    phantoms<{ T: T; P: P }>(),
    {
      "": { toSchema, context, applied },
      fill: <A extends Partial<Record<P, string | number>>>(values: A) => {
        return Ty<T, Exclude<P, keyof A>>(toSchema, context, { ...applied, ...values })
      },
      isRoot(): this is RootTy {
        return false
      },
    },
  )
}

export interface Ty<T = any, P extends keyof any = keyof any> {
  <P2 extends Array<keyof any>>(
    template: TemplateStringsArray,
    ...placeheld: P2
  ): Ty<T, P | P2[number]>
  T: T
  P: P
  "": {
    toSchema: ToSchema
    context: Array<Context>
    applied: Applied
  }
  fill: <A extends Partial<Record<P, number | string>>>(values: A) => Ty<T, Exclude<P, keyof A>>
  isRoot(): this is RootTy
}

export type ToSchema = (ref: Ref) => Schema
export type Schema = Record<string, unknown>

export interface Context {
  template: TemplateStringsArray
  placeheld: Array<keyof any>
}

export type Applied = Record<keyof any, number | string>
