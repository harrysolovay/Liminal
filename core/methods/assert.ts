import { assert as assert_, assertArrayIncludes, assertEquals } from "@std/assert"
import type { Diagnostic } from "../Diagnostic.ts"
import type { AnyType } from "../Type.ts"
import { TypeVisitor } from "../TypeVisitor.ts"

export function assert<T>(this: AnyType<T>, value: unknown): asserts value is T {
  const diagnostics: Array<Diagnostic> = []
  visit(new AssertContext(diagnostics, "root", value), this)
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}

export class AssertContext {
  constructor(
    readonly diagnostics: Array<Diagnostic>,
    readonly path: string,
    readonly value: unknown,
    readonly junction?: number | string,
  ) {}

  descend = (value: unknown, junction?: number | string): AssertContext =>
    new AssertContext(this.diagnostics, this.path, value, junction)
}

const visit = TypeVisitor<AssertContext, void>({
  hook(next, ctx, type) {
    const path = ctx.junction ? ctx.path : `${ctx.path}["${ctx.junction}"]`
    try {
      next(ctx, type)
    } catch (exception: unknown) {
      ctx.diagnostics.push({
        type,
        path,
        exception,
        value: ctx.value,
      })
    }
  },
  null({ value }) {
    assert_(value === null)
  },
  boolean({ value }) {
    assert_(typeof value === "boolean")
  },
  integer({ value }) {
    assert_(Number.isInteger(value))
  },
  number({ value }) {
    assert_(typeof value === "number")
  },
  string({ value }) {
    assert_(typeof value === "string")
  },
  const({ value }, _1, _2, cmpValue) {
    assertEquals(value, cmpValue)
  },
  enum({ value }, _1, ...values) {
    assert_(typeof value === "string")
    assertArrayIncludes(values, [value])
  },
  array(ctx, _1, element) {
    assert_(Array.isArray(ctx.value))
    ctx.value.forEach((value, i) => visit(ctx.descend(value, i), element))
  },
  object(ctx, _1, fields) {
    const keys = Object.keys(fields).toSorted()
    const { value } = ctx
    assert_(typeof value === "object" && value !== null)
    assertEquals(keys, Object.keys(value).toSorted())
    keys.forEach((key) => visit(ctx.descend(value[key as never], key), fields[key]!))
  },
  transform(ctx, _1, from) {
    visit(ctx, from)
  },
  union(ctx, _1, ...members) {
    const match = matchUnionMember(members, ctx.value)
    assert_(match)
    visit(ctx, match)
  },
  f(ctx, _1, get) {
    visit(ctx, get())
  },
})

export function matchUnionMember(members: Array<AnyType>, value: unknown): AnyType | void {
  const queue = [...members]
  while (queue.length) {
    const current = queue.pop()!
    const diagnostics: Array<Diagnostic> = []
    visit(new AssertContext(diagnostics, "", value), current)
    if (diagnostics.length) {
      continue
    }
    return current
  }
}
