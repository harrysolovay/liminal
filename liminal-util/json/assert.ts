import { assert as assert_ } from "./assert.ts"
import type { Value } from "./Value.ts"

export function assert(value: unknown): asserts value is Value {
  if (value === null) {
    return
  } else if (Array.isArray(value)) {
    value.every(assert)
  } else if (typeof value === "object") {
    Object.values(value).every(assert)
  } else {
    assert_(typeof value === "string" || typeof value === "number" || typeof value === "boolean")
  }
}
