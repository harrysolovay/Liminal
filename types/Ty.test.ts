import { assertEquals } from "@std/assert"
import { assertTySnapshot } from "test_util"
import { SubschemaFactory, Ty } from "./Ty.ts"

export const any = Ty<unknown, never, true>(() => ({
  type: ["string", "number", "object", "array", "boolean", "null"],
}), true)

Deno.test("Ty", async (t) => {
  await assertTySnapshot(t, any)
})

const subschema = SubschemaFactory({})

Deno.test("context chaining", () => {
  const a = any`A.`
  assertEquals(subschema(a).description, "A.")

  const b = a`B.`
  assertEquals(subschema(b).description, "B. A.")

  const c = b`C.`
  assertEquals(subschema(c).description, "C. B. A.")
})

Deno.test("numeric parameter key", () => {
  const t = any`Arg: ${1}.`
  assertEquals(subschema(t.fill({ 1: "value" })).description, "Arg: value.")
})

Deno.test("string parameter key", () => {
  const t = any`Arg: ${"P"}.`
  assertEquals(subschema(t.fill({ P: "value" })).description, "Arg: value.")
})

Deno.test("symbol parameter key", () => {
  const sym = Symbol()
  const t = any`Arg: ${sym}.`
  assertEquals(subschema(t.fill({ [sym]: "value" })).description, "Arg: value.")
})

Deno.test("parameterized with chaining", () => {
  const a = any`a: ${"A"}.`
  const b = a`b: ${"B"}.`
  const c = b`c: ${"C"}.`
  const d = c.fill({ A: "A", B: "B", C: "C" })
  assertEquals(subschema(d).description, "c: C. b: B. a: A.")
})
