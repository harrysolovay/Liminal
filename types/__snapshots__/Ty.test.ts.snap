export const snapshot = {};

snapshot[`Ty 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      type: [
        "string",
        "number",
        "object",
        "array",
        "boolean",
        "null",
      ],
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Ty 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "One.",
      type: [
        "string",
        "number",
        "object",
        "array",
        "boolean",
        "null",
      ],
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Ty 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      description: "Two. One.",
      type: [
        "string",
        "number",
        "object",
        "array",
        "boolean",
        "null",
      ],
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
